### ISSUE TITLE
[RFC][FSDP] Rework `FlatParameter`, `FlattenParamsWrapper`

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#76501]
### TL;DR (User Visible Changes)
- The original model parameter variables are kept as `nn.Parameter`s instead of changed to `Tensor`s.
- The original model parameter variables' `.grad` variable accesses the sharded gradient.
- The user may pass the original model parameters to the optimizer instead of the `FlatParameter`s, enabling multiple parameter groups.
- `FullyShardedDataParallel.parameters()` returns the original model parameter variables, excluding any `FlatParameter`s.
- `FullyShardedDataParallel.flat_parameters()` (new API) returns the `FlatParameter`s (matching the semantics of the old `FullyShardedDataParallel.parameters()`).

### Problem & Motivation
Currently, `FullyShardedDataParallel` (FSDP) [replaces the original model parameter variables with `Tensor`s that are views into a `FlatParameter`](https://github.com/pytorch/pytorch/blob/ea1901693e92266a61950ed61bfe8f6068a81b98/torch/distributed/fsdp/flatten_params_wrapper.py#L400-L405). This has some implications:
- The original model parameters' `.grad`s are not set, meaning that they cannot be optimized separately (which is why we cannot support multiple parameter groups).
- The original model parameter variables are invisible to `FSDP.parameters()` since they are not registered as `nn.Parameter`s.
    - As a result, `summon_full_params()` temporarily re-registers the original model parameter variables as `nn.Parameter`s in a context manager to be able to use `nn.Module.parameters()`.
- The views are used to perform the `forward()` computation, and the gradient computation relies on `torch.split()`'s adjoint to propagate the gradient from the views to the `FlatParameter`'s `.grad`.

The primary goal of this RFC is to set each original model parameter's `.grad` to enable multiple parameter groups, but along the way, it can remove the logic like re-registering the model parameters as `nn.Parameter`s just for `summon_full_params()`.

### Approach
- We keep the original model parameter variables as `nn.Parameter`s and manipulate their storage and `.grad` variable manually. 
    - Like before, we set each parameter's storage to an offset in the `FlatParameter`'s storage.
        - However, this operation is no longer tracked by the autograd engine.
    - We additionally set each parameter's `.grad` to an offset in a `FlatParameter`'s `.grad`.
        - In the pre-backward hook, we manually allocate the `FlatParameter`'s `.grad` if needed (this should happen once) and then point each parameter's `.grad` into the `FlatParameter`'s `.grad`.
- **The `FlatParameter` is simply a container for the contiguous parameter and gradient memory** (enabling efficient sharding and communication). However, we ensure that both the `FlatParameter`'s `.grad` and each parameter's `.grad` are correct, so the user may choose to pass either `FlatParameter`s or the original parameters into an optimizer.
    - This helps preserve backward compatibility.
    - The autograd engine propagates gradients through the original parameters with no awareness of the `FlatParameter`s. It just happens to be that the original parameters' `.grad` have storage that is part of a contiguous memory block saved in a `FlatParameter`'s `.grad`.
- Because the autograd engine no longer propagates gradients through the `FlatParameter`s and then through the views (via `torch.split()`'s adjoint), we must change the way we register post-backward hooks, which previously were registered on each `FlatParameter`'s `AccumulateGrad` object.
    - We must now register hooks on each **parameter**'s `AccumulateGrad` object. (Empirically, I have found that the `AccumulateGrad` object must be saved or else a new `AccumulateGrad` object may be used, in which case the registered hook is not run).
    - To still ensure that the existing post-backward logic only runs once the gradients for all parameters comprising a `FlatParameter` have been computed, we may do the following:
        - We use a data structure, e.g. `_param_to_is_grad_ready`, initialized to `None`.
        - In the first pre-backward hook run (indicated by the structure being `None`), we initialize it by iterating over the module's parameters and adding an entry mapping each parameter with `requires_grad=True` to `False`, indicating that its gradient is not yet ready.
        - In the post-backward hook, we set the parameter's `_param_to_is_grad_ready[param] = True` and check `all(_param_to_is_grad_ready.values())`. If this is `True`, then all gradients are ready, and we may run the existing post-backward hook logic (e.g. to reduce-scatter the `FlatParameter`'s `.grad`).
    - Using `_param_to_is_grad_ready` should enable individual parameters to be frozen dynamically.
- Because the original model parameters are kept as `nn.Parameter`s, they are visible to `nn.Module.parameters()`.
    - To preserve the existing usages of `nn.Module.parameters()` that expect `FlatParameter`s only, we may introduce a new API `flat_parameters()` and `named_flat_parameters()`.
    - `super().named_parameters()` will return a mix of both `FlatParameter`s and original model parameters, so we can override `named_parameters()` to exclude `FlatParameters` and have `named_flat_parameters()` only include `FlatParameter`s. 

### Prototype
The most uncertain parts are surrounding autograd and the hook usage. To reduce this uncertainty, I have put together a prototype exercising some of the techniques in [Approach](###Approach). See the script [here](https://github.com/awgu/views/blob/master/views.py).

### Feedback
- How do we handle the case when a gradient is ready multiple times? E.g. if the same parameter is involved multiple times in the computational graph?
- How do we deal with non in-place modifications to the original model parameters? (We need to copy these changes back to the `FlatParameter` storage.)

cc @zhaojuanmao @mrshenli @rohan-varma

**Comment by zhaojuanmao** [comment]
Thanks for the proposal and the script. The workflow looks great in the script!

Regarding current script, here are a few feedback: 
1. since original parameters are restored as module's parameters outside computation, although we make the original parameters point to the underneath flat_param storage, could we still double check the peak memory impact during implementation just in case? 
2. in real FSDP implementation, the flat_param is sharded and storage is completely changed outside computation, so we still need to point the original parameters to sharded flat_param by offsets of original parameters in the sharded flat_param, right? 
3. if we want to support the mode to allow users to access the original parameters directly outside computation, we can add a flag in the main API, so that multiple param groups and fine tuning use cases can just simply set this flag and turn on the code paths discussed here
4. for the mode mentioned above, possibly we need some changes for model and optimize state_dict APIs as well. Not sure, possibly not major changes

**Comment by awgu** [comment]
Commenting to say that this RFC needs to be updated with new developments and discussion. I will update it some time after the branch cut (5/19).

**Comment by awgu** [comment]
Closing due to more internal discussion.

**Comment by wukong1992** [comment]
how is the rfc going？
